
### DEFAULT CONFIG FILE

import os

configfile: os.path.join(workflow.basedir, '../', 'config', 'config.yaml')

### from config files
#  input as csv
CSV = config['input']
OUTPUT = config['output']
THREADS = config['threads']

# snakemake params 
BigJobMem = config["BigJobMem"]
BigJobCpu = config["BigJobCpu"]
SmallJobMem = config["SmallJobMem"]
SmallJobCpu = config["SmallJobCpu"]

SmallTime = config["SmallTime"]
BigTime = config["BigTime"]
MediumTime = config["MediumTime"]

# plassembler DB
PlassemblerDatabase = config["plassemblerDatabase"]

# LR Only flag 
LR_ONLY = config['long']

# min chrom length for 
MIN_LENGTH = config['min_length']
MIN_QUALITY = config['min_quality']

POLCA_FLAG = config['polca']

MEDAKA_MODEL = config['medakaModel']
FLYE_MODEL = config['flyeModel']

# define functions
# get long reads
def get_input_lr_fastqs(wildcards):
    return dictReads[wildcards.sample]["LR"]


### Include Directories
include: "rules/directories.smk"

# Parse the samples and read files
include: "rules/samples.smk"
dictReads = parseSamples(CSV, LR_ONLY)
SAMPLES = list(dictReads.keys())

wildcard_constraints:
    sample= '|'.join([re.escape(x) for x in SAMPLES])


# Import rules and functions
include: "rules/targets.smk"
include: "rules/aggregate.smk"
include: "rules/qc.smk"
include: "rules/assemble.smk"
include: "rules/assembly_statistics.smk"
include: "rules/extract_fastas.smk"

# checkpoint here for completeness
include: "rules/long_read_polish.smk"
include: "rules/long_read_incomplete.smk"



#  polish the assemblies
if LR_ONLY == False:
    include: "rules/short_read_polish.smk"
    include: "rules/short_read_polish_incomplete.smk"
    if POLCA_FLAG == True:
        include: "rules/short_read_polca.smk"



if LR_ONLY == False:
    include: "rules/plassembler.smk"
    include: "rules/combine_plassembler_info.smk"

### need to get final chromosome and plasmids

rule all:
    input:
        TargetFiles
